Add key transform functions for massive rate limit improvements

v-stickykeys opened the pull request:
This is simply an extension of the work @obo20 did in this PR https://github.com/ipfs/go-ds-s3/pull/195

It incorporates the suggestion made by @whyrusleeping https://github.com/ipfs/go-ds-s3/pull/195#issuecomment-900466233 to allow a key transform function to be specified such that developers can decide the layout of their data on S3.

In particular, I included the [`next-to-last/2`](https://github.com/ipfs/js-datastore-core/blob/master/src/shard-readme.js) method used in the [js-datastore-s3](https://github.com/ipfs/js-datastore-s3) to make it possible to use go-ipfs or js-ipfs over the same datastore.

v-stickykeys commented:
I wasn't able to test this unless the module name matched my repo name so that is one thing I will change (may need some guidance to figure out how to do this properly as I'm new to go!)

BigLep commented:
@MichaelMure : are you able to review this?

obo20 commented:
From conversations with @whyrusleeping , it's my understanding that the next-to-last/2 method (and other methods of filesystem sharding) were originally implemented in order to better scale linux based filesystems.

@achingbrain would you be able to share the reasons for sharding things out in the way you did in js-datastore-s3? Did you run into similar scalability issues or were you just trying to follow the existing filesystem sharding patterns of go-ipfs?

Also, @aschmahmann / @BigLep as we move towards go-ipfs 0.12.0, do any considerations need to be taken with the desire to "Switch to raw multihashes in datastores (including datastore version upgrade)" in the 0.12.0 upgrade? If possible, I'd love to make sure any possible compatibility issues are addressed pre-emptively here.

Stebalien commented:
@obo20 see https://github.com/ipfs/go-ds-s3/issues/105 and https://github.com/ipfs/go-ds-s3/issues/199 (specifically the second one).

Stebalien commented:
Oh, wait, sorry. You already know that.

For sharding _blocks_, "next to last" is probably fine. Although I'd personally just use a cheap hash function.

achingbrain commented:
> would you be able to share the reasons for sharding things out in the way you did in js-datastore-s3? Did you run into similar scalability issues or were you just trying to follow the existing filesystem sharding patterns of go-ipfs?

@jacobheun did the initial implementation so perhaps he can speak to the reason, but I get the feeling it was just following the principles set out by other datastore implementations.

jacobheun commented:
> @jacobheun did the initial implementation so perhaps he can speak to the reason, but I get the feeling it was just following the principles set out by other datastore implementations.

This is correct, it was just following the patterns of the other JS datastore implementations at the time (2018), it's just a lightweight interface to s3. @achingbrain I believe the sharding behavior changed in https://github.com/ipfs/js-datastore-s3/pull/30, but it's not managed by the s3 datastore itself.

v-stickykeys commented:
@MichaelMure I believe this is ready for review. PTAL when you get the chance!

v-stickykeys commented:
@BigLep given no word so far from @MichaelMure is there someone else who might be available to review this?

BigLep commented:
Similar response as in https://github.com/ipfs/go-ds-s3/pull/195#issuecomment-1013311478

Core IPFS maintainers aren't planning to take this on as we aren't using this ourselves. We need to figure out a way to setup and denote community membership for this and other datastore repos. We're game to setup others as the maintainers here.

Maybe pair up with @obo20  and review each other's code (@obo20 has #195 open) or ask for other volunteers in the ipfs-operators Slack channel?

guseggert commented on s3.go:
```diff
@@ -56,6 +56,22 @@ type Config struct {
+}
+
+var KeyTransforms = map[string]func(ds.Key) string{
+	"default": func(k ds.Key) string {
```
Calling this "default" isn't very descriptive, how about "identity" or something like that?

guseggert commented on s3.go:
```diff
@@ -56,6 +56,22 @@ type Config struct {
+	"suffix": func(k ds.Key) string {
+		return k.String() + "/data"
+	},
+	"next-to-last/2": func(k ds.Key) string {
+		s := k.String()
+		offset := 1
+		start := len(s) - 2 - offset
+		return s[start:start+2] + "/" + s
+	},
```
I think there are some problems here

```
datastore.NewKey("a") -> panic
datastore.NewKey("ab") -> /a//ab
datastore.NewKey("abc") -> ab//abc
datastore.NewKey("abcd") -> bc//abcd
```

* Panics on keys w/ len == 1, should return an error instead
* Keys w/ len == 2 have a slash in their prefix (should probably return an error instead?)
* Keys w/ len > 2 have two slashes instead of one

A small unit test would be nice here.

v-stickykeys replied:
Thanks for this @guseggert ! In practice, I am not seeing the double slash issue. Also just running through the logic I don't see how that would be possible but I'm new to Go 🙈.

However for keys of len 2 or smaller I agree this should be handled differently.
For keys of length 3 or more it still seems correct given a param of 2.

guseggert replied:
All I did was add this test to print out the results, should help you reproduce:

```go
func TestTransforms(t *testing.T) {
	transform := KeyTransforms["next-to-last/2"]
	keys := []string{"ab", "abc", "abcd"}
	for _, k := range keys {
		fmt.Printf("datastore.NewKey(\"%s\") -> %s\n", k, transform(datastore.NewKey(k)))
	}
}
```


Full test:


```go
func TestTransforms(t *testing.T) {
	cases := []struct {
		transform     string
		key           string
		expectedS3Key string
	}{
		{
			transform:     "next-to-last/2",
			key:           "ab",
			expectedS3Key: "ab/ab",
		},
		{
			transform:     "next-to-last/2",
			key:           "abc",
			expectedS3Key: "ab/abc",
		},
		{
			transform:     "next-to-last/2",
			key:           "abcd",
			expectedS3Key: "ab/abcd",
		},
	}
	for _, c := range cases {
		t.Run(fmt.Sprintf("%s %s", c.transform, c.key), func(t *testing.T) {
			transform := KeyTransforms[c.transform]
			dsKey := datastore.NewKey(c.key)
			s3Key := transform(dsKey)
			if s3Key != c.expectedS3Key {
				t.Fatalf("expected %s, got %s", c.expectedS3Key, s3Key)
			}
		})
	}
}
```

output:

```
go test -v -run TestTransforms\$ . 
=== RUN   TestTransforms
=== RUN   TestTransforms/next-to-last/2_ab
    s3_test.go:105: expected ab/ab, got /a//ab
=== RUN   TestTransforms/next-to-last/2_abc
    s3_test.go:105: expected ab/abc, got ab//abc
=== RUN   TestTransforms/next-to-last/2_abcd
    s3_test.go:105: expected ab/abcd, got bc//abcd
--- FAIL: TestTransforms (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_ab (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_abc (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_abcd (0.00s)
FAIL
FAIL	github.com/3box/go-ds-s3	0.004s
FAIL
```

guseggert commented on s3_test.go:
```diff
@@ -17,20 +19,26 @@ func TestSuiteLocalS3(t *testing.T) {
 		t.Skipf("skipping test suit; LOCAL_S3 is not set.")
 	}
 
+  localBucketName, localBucketNameSet := os.LookupEnv("LOCAL_BUCKET_NAME")
+  if !localBucketNameSet {
+    localBucketName = fmt.Sprintf("localbucketname%d", time.Now().UnixNano())
+	}
```
go fmt?

guseggert commented on s3_test.go:
```diff
@@ -17,20 +19,26 @@ func TestSuiteLocalS3(t *testing.T) {
 		Region:         "local",
 		AccessKey:      "test",
 		SecretKey:      "testdslocal",
+		KeyTransform:   "default",
```
Could you add tests for the other transforms?

guseggert commented on plugin/main/main.go:
```diff
@@ -1,7 +1,7 @@
 
 import (
-	plugin "github.com/ipfs/go-ds-s3/plugin"
+	plugin "github.com/3box/go-ds-s3/plugin"
```
```suggestion
	plugin "github.com/ipfs/go-ds-s3/plugin"
```

guseggert commented on plugin/s3ds.go:
```diff
@@ -3,7 +3,7 @@ package plugin
 	"fmt"
 
-	s3ds "github.com/ipfs/go-ds-s3"
+	s3ds "github.com/3box/go-ds-s3"
```
```suggestion
	s3ds "github.com/ipfs/go-ds-s3"
```

guseggert commented on plugin/s3ds_test.go:
```diff
@@ -4,7 +4,7 @@ import (
 	"testing"
 
-	s3ds "github.com/ipfs/go-ds-s3"
+	s3ds "github.com/3box/go-ds-s3"
```
```suggestion
	s3ds "github.com/ipfs/go-ds-s3"
```

guseggert commented on README.md:
```diff
@@ -81,6 +82,18 @@ The config file should include the following:
+`suffix`
+- Shards by storing block data at a key with a `data` suffix. E.g. `CIQJ7IHPGOFUJT5UMXIW6CUDSNH6AVKMEOXI3UM3VLYJRZUISUMGCXQ/data`
+
+`next-to-last/2`
+- Shards by storing block data based on the second to last 2 characters of its key. E.g. `CX/CIQJ7IHPGOFUJT5UMXIW6CUDSNH6AVKMEOXI3UM3VLYJRZUISUMGCXQ`
```
```suggestion
`next-to-last/2`
- Shards by storing block data based on the second to last 2 characters of its key. E.g. `CX/CIQJ7IHPGOFUJT5UMXIW6CUDSNH6AVKMEOXI3UM3VLYJRZUISUMGCXQ`
- This is for compatibility with [js-datastore-s3](https://github.com/ipfs/js-datastore-s3)
```

BigLep commented:
@v-stickykeys : are you going to be able to incorporate the feedback?  We'd like to support getting this landed.

v-stickykeys commented:
Hey @BigLep the short answer is yes but honestly it hasn't been the top of my priority list and is now more of a weekend project. My PR as it stands meets our needs but I do have another branch with improvements addressing the review which I can try to get wrapped and pushed up in the next week.

BigLep commented:
@v-stickykeys : just checking in.  Maintainers can take a look if you post an new revision with incorporated comments.

koxon commented:
Hi guys. This would be a shame not to merge this. 

I wrote the [IPFS on AWS Fargate blog](https://aws.amazon.com/blogs/containers/deploying-ipfs-cluster-using-aws-fargate-and-amazon-efs-one-zone/) for AWS and I'm currently looking at making this S3 plugin work with this architecture.

S3 Sharding is key to make this a viable option compared to traditional drives. 
With IPFS we're looking for high throughput. Low latency data access is less important. 

Referencing [this post](https://github.com/ipfs/go-ds-s3/issues/105) from @mikeal if we can achieve 40GB/s throughput per bucket we could run large cluster at a fraction of the cost. We can use multiple buckets in multiple regions if we need to scale. 

I will try this branch and I'm willing to help move this forward.

@guseggert you've submitted multiple feedback. What is your assessment of the situation?
Looks like there is minor work left to finalize.

@v-stickykeys are you still around?

hsanjuan commented:
@koxon sorry that no one answered this... are you still interested in picking this up?

Right now I don't see any upsides in letting users configure the number of prefixes between next-to-last-2,3 or whatever... might as well set it to a number that 128 or a number that maximizes the usage of S3 and leave it there. Unless there is a good reason to use a worse number.
